src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStackFrame.cs

```diff
@@ -94,16 +94,16 @@ public JsonPropertyInfo GetPolymorphicJsonPropertyInfo()
         }
 
         /// <summary>
-        /// Initializes the state for polymorphic or re-entry cases.
+        /// Initializes the state for polymorphic cases.
         /// </summary>
-        public JsonConverter InitializeReEntry(Type type, JsonSerializerOptions options, string? propertyName = null)
+        public JsonConverter InitializeReEntry(Type type, JsonSerializerOptions options)
         {
-            JsonClassInfo classInfo = options.GetOrAddClass(type);
+            if (PolymorphicJsonPropertyInfo?.RuntimePropertyType != type)
```

Comment:
Would it be beneficial (or even possible) to move this check into the caller, here?
https://github.com/dotnet/runtime/blob/746ca52cf3405d9faeb3d2644b56e3761b375002/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs#L337

Also, please add a comment explaining what this check is doing and when it will be false such that we can skip calling `options.GetOrAddClass` (something inspired by your PR description would be a good starting point).

Reply:
> Would it be beneficial (or even possible) to move this check into the caller (looks like there is only one callsite), here?

It is possible, but I think it breaks some encapsulation by that method. Also, the `PolymorphicJsonPropertyInfo` field should be changed from public to private since it already exposed by `GetPolymorphicJsonPropertyInfo()` which should normally be used to obtain that.

> Also, please add a comment..

Yep. Will do